proposals/ref-struct-interfaces.md

```diff
@@ -277,6 +285,418 @@ Specifically by using the `CorGenericParamAttr.gpAllowByRefLike(0x0020)` or `Sys
+    static void Test<TEnumerable, TEnumerator>(TEnumerable t)
+        where TEnumerable : IGetEnumerator<TEnumerator>
+        where TEnumerator : ICustomEnumerator, System.IDisposable, allows ref struct 
+    {
+        foreach (var i in t)
+        {
+        } // System.IDisposable.Dispose()
+    }
+}
+```
+
+Note that a pattern `Dispose` method will not be recognized on a type parameter that `allows ref struct` because
+an interface (and this is the only place where we could possibly look for a pattern) is not a ref struct.
+Also, since runtime doesn't provide a way to check whether at runtime a type parameter that `allows ref struct`
+implements `IDisposable` interface, a type parameter enumerator that `allows ref struct` will be disallowed,
+unless `IDisposable` is in its effective interfaces set.
+```csharp
+interface ICustomEnumerator
+{
+    int Current {get;}
+    bool MoveNext();
+}
+
+interface IMyDisposable
+{
+    void Dispose();
+}
+
+interface IGetEnumerator<TEnumerator> where TEnumerator : allows ref struct 
+{
+    TEnumerator GetEnumerator();
+}
+
+class C
+{
+    static void Test<TEnumerable, TEnumerator>(TEnumerable t)
+        where TEnumerable : IGetEnumerator<TEnumerator>
+        where TEnumerator : ICustomEnumerator, IMyDisposable, allows ref struct 
+    {
+        // error CS9507: foreach statement cannot operate on enumerators of type 'TEnumerator'
+        //               because it is a type parameter that allows ref struct and
+        //               it is not known at compile time to implement IDisposable.
+        foreach (var i in t)
+        {
+        }
+    }
+}
+```
+
+### `await foreach` statement
+
+An `await foreach` statement will recognize and use implementation of ```IAsyncEnumerable<T>``` interface when collection is a ref struct.
+```csharp
+ref struct S : IAsyncEnumerable<int>
+{
+    IAsyncEnumerator<int> IAsyncEnumerable<int>.GetAsyncEnumerator(CancellationToken token) {...}
+}
+
+class C
+{
+    static async Task Main()
+    {
+        await foreach (var i in new S()) // S.IAsyncEnumerable<int>.GetAsyncEnumerator
+        {
+        }
+    }
+}
+```
+
+A pattern `GetAsyncEnumerator` method will be recognized on a type parameter that `allows ref struct` as it is recognized on
+type parameters without that constraint today.
+
+```csharp
+interface IMyAsyncEnumerable<T>
+{
+    IAsyncEnumerator<int> GetAsyncEnumerator(CancellationToken cancellationToken = default);
+}
+
+class C
+{
+    static async Task Test<T>() where T : IMyAsyncEnumerable<int>, allows ref struct
+    {
+        await foreach (var i in default(T)) // IMyAsyncEnumerable<int>.GetAsyncEnumerator
+        {
+        }
+    }
+}
+```
+
+An `await foreach` statement will recognize and use implementation of ```IAsyncEnumerable<T>``` interface when collection is a type parameter that 
+`allows ref strict`, there is no `GetAsyncEnumerator` pattern match, and ```IAsyncEnumerable<T>``` is in type parameter's effective interfaces set.
```

Comment:
```suggestion
`allows ref struct`, there is no `GetAsyncEnumerator` pattern match, and ```IAsyncEnumerable<T>``` is in type parameter's effective interfaces set.
```

Reply:
> An `await foreach` statement will recognize and use implementation of ```IAsyncEnumerable<T>``` interface when collection is a type parameter that 
> `allows ref strict`, there is no `GetAsyncEnumerator` pattern match, and ```IAsyncEnumerable<T>``` is in type parameter's effective interfaces set.

Currently, `IAsyncEnumerable<T>` is used even if there is a `GetAsyncEnumerator` pattern (available through an interface) - see [sharplab.io](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIpwCwG5lqZwB0AKgBYBOApgIbACWAdgOb4oZYCs7yzALtUoAzWgGNq6AJIBZAJ4BBAM5ymYgKJMArgFshtAEYAbanAA8pAHzIA3snQPpy1Ru17KtfgHtKZgZfQAcWp+ZzVNXX1vSgAKAGFaNWojI08GLyZSLwBraiZ0MUSJFLSMrNz8gF50YGpRLSN+AEp2AF9kAhh0ONt7RywANkwADkwh0molfgtLGKb0AHdyIUlSdBBpeTDXSI9jUz8mfksAGicVcLd9fcPjvoc7JEdnzABOdGEfOjFydBiAN1olHQDBB+Vq9UaMVITXmAHo4ecXBF3IYTLdLMRgqELjtUdF7i9Hi8SURiHEMkovCZiAB1SgMQQxBgtQnPdpPRwc1pAA===).

Reply:
> > An `await foreach` statement will recognize and use implementation of `IAsyncEnumerable<T>` interface when collection is a type parameter that
> > `allows ref strict`, there is no `GetAsyncEnumerator` pattern match, and `IAsyncEnumerable<T>` is in type parameter's effective interfaces set.
> 
> Currently, `IAsyncEnumerable<T>` is used even if there is a `GetAsyncEnumerator` pattern (available through an interface) - see [sharplab.io](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIpwCwG5lqZwB0AKgBYBOApgIbACWAdgOb4oZYCs7yzALtUoAzWgGNq6AJIBZAJ4BBAM5ymYgKJMArgFshtAEYAbanAA8pAHzIA3snQPpy1Ru17KtfgHtKZgZfQAcWp+ZzVNXX1vSgAKAGFaNWojI08GLyZSLwBraiZ0MUSJFLSMrNz8gF50YGpRLSN+AEp2AF9kAhh0ONt7RywANkwADkwh0molfgtLGKb0AHdyIUlSdBBpeTDXSI9jUz8mfksAGicVcLd9fcPjvoc7JEdnzABOdGEfOjFydBiAN1olHQDBB+Vq9UaMVITXmAHo4ecXBF3IYTLdLMRgqELjtUdF7i9Hi8SURiHEMkovCZiAB1SgMQQxBgtQnPdpPRwc1pAA===).

I can see how this could be confusing, but there is no pattern match in this case due to an ambiguity.